✨ Add support for EKS Provisioned Control Plane

cespo opened the pull request:
<!-- Thanks for this PR! If this is your first PR please read the [contributing guide](../CONTRIBUTING.md) -->

**What type of PR is this?**

/kind feature
/kind api-change

**What this PR does / why we need it**:

This PR adds support for the EKS Provisioned Control Plane feature, which allows cluster administrators to select from a set of scaling tiers for predictable, high performance from the cluster's control plane.

Amazon EKS offers two modes of operations for the control plane: Standard mode (default, with automatic scaling) and Provisioned mode (pre-allocated capacity). This implementation enables users to opt-in to Provisioned mode and select from the available scaling tiers: `standard`, `tier-xl`, `tier-2xl`, or `tier-4xl`.

**Changes:**
- Added `ControlPlaneScalingConfig` field to `AWSManagedControlPlane` API (v1beta2)
- Added `ControlPlaneScalingTier` enum type with kubebuilder validation
- Implemented converters between CAPA and AWS SDK types
- Updated cluster creation to include scaling configuration
- Added reconciliation logic for updating scaling configuration on existing clusters
- Updated v1beta1 conversion functions for backward compatibility
- Upgraded AWS SDK for EKS from v1.64.0 to v1.75.1 to support the new API

**Usage example:**
```yaml
apiVersion: controlplane.cluster.x-k8s.io/v1beta2
kind: AWSManagedControlPlane
metadata:
  name: my-cluster-control-plane
spec:
  region: us-west-2
  version: "1.28"
  controlPlaneScalingConfig:
    tier: tier-2xl  # Options: standard, tier-xl, tier-2xl, tier-4xl
```

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:

Fixes #5771

**Special notes for your reviewer**:

- The new field is optional and maintains backward compatibility - existing clusters continue to use standard mode
- The AWS SDK upgrade was necessary as `ControlPlaneScalingConfig` support was added in a recent SDK version
- Conversion functions ensure the field is preserved when converting between v1beta1 and v1beta2
- The reconciliation logic handles both cluster creation and updates to the scaling tier
- CRD changes are included with proper validation (enum constraint on tier values)

**Checklist**:

- [x] squashed commits
- [x] includes documentation (inline comments and usage example)
- [x] includes emoji in title (:sparkles:)
- [x] adds unit tests
- [x] adds or updates e2e tests

**Release note**:

```release-note
Add support for EKS Provisioned Control Plane, allowing users to configure control plane scaling tiers (standard, tier-xl, tier-2xl, tier-4xl) for predictable high performance. This feature is optional via the new `controlPlaneScalingConfig` field in `AWSManagedControlPlane`.
```

serngawy reviewed (commented):
/ok-to-test

cespo commented:
/retest

faiq commented:
Can you add an e2e test as well?

cespo commented:
/retest

cespo commented:
/retest pull-cluster-api-provider-aws-lint-api

cespo commented:
I've verified `make lint-api` passes locally with 0 issues. This appears to be a CI infrastructure problem.

cespo commented:
/retest

cespo commented:
The lint-api job is consistently failing during setup (unable to clone golangci-lint). This appears to be a CI infrastructure issue, not related to my changes. cc @faiq @serngawy @ncdc

cespo commented:
> @cespo please rebase. we fixed the issue

Done

cespo commented:
bump 
@faiq @serngawy @fiunchinho @dlipovetsky any chance to get this reviewed?

faiq commented on controlplane/eks/api/v1beta2/types.go:
```diff
@@ -259,6 +259,32 @@ func (e UpgradePolicy) String() string {
+	// Valid values are: standard, tier-xl, tier-2xl, tier-4xl
+	// +optional
+	// +kubebuilder:validation:Enum=standard;tier-xl;tier-2xl;tier-4xl
+	Tier *ControlPlaneScalingTier `json:"tier,omitempty"`
```
we should really document what happens if you don't fill this optional field out. what is the default? assuming its standard?

faiq replied:
Can we document that somewhere in the code

cespo commented:
All comments have been addressed. I would appreciate a review please

aaresh-sharma commented:
@fiunchinho @dlipovetsky Any progress on the reviews?

cespo commented:
@fiunchinho @dlipovetsky any chance this change could be approved and merged in?

richardcase commented on controlplane/eks/api/v1beta2/types.go:
```diff
@@ -293,6 +293,34 @@ func (e UpgradePolicy) String() string {
+	// When this field is not set, EKS defaults to the standard tier (automatic control plane scaling).
+	// +optional
+	// +kubebuilder:validation:Enum=standard;tier-xl;tier-2xl;tier-4xl
+	Tier *ControlPlaneScalingTier `json:"tier,omitempty"`
```
I think it's probably worth adding the default using the kubebuilder annotations and changing it from a pointer. The reason is i feel that its good to be explicit about the default value if we can, instead of it just being a comment.

Although i'm not strongly opinionated on this 😄

cespo replied:
Yeah it makes sense. I'll change Tier to a value type

richardcase commented on pkg/cloud/services/eks/cluster.go:
```diff
@@ -838,6 +846,26 @@ func (s *Service) reconcileUpgradePolicy(upgradePolicy *ekstypes.UpgradePolicyRe
+	}
+
+	// If the cluster doesn't have a scaling config yet, or if it differs from the spec, update it
+	desiredConfig := converters.ControlPlaneScalingConfigToSDK(s.scope.ControlPlane.Spec.ControlPlaneScalingConfig)
```
So you can change the scaling tier after the cluster has been created?

richardcase replied:
Looking at the docs, yes you can. Thats good.

cespo replied:
Yes you can. We updated a bunch of clusters in place. We had cases of clusters going to provisioned CP and back to default without issues

richardcase commented on pkg/cloud/services/eks/cluster.go:
```diff
@@ -838,6 +846,26 @@ func (s *Service) reconcileUpgradePolicy(upgradePolicy *ekstypes.UpgradePolicyRe
+	}
+
+	// If current config is nil or tier differs, update is needed
+	if scalingConfig == nil || scalingConfig.Tier != desiredConfig.Tier {
```
Its probably worth adding a `Equals` to `ControlPlaneScalingConfig` so that it can be used with `cmp.Equals` (which it can be without that Equals function but it makes it more explicit).

cespo replied:
I will add an Equal method on `*ControlPlaneScalingConfig` and will refactor `reconcileScalingConfig` to use it

richardcase commented on test/e2e/suites/managed/eks_scaling_config_test.go:
```diff
@@ -0,0 +1,146 @@
+// +build e2e
+
+/*
+Copyright 2025 The Kubernetes Authors.
```
```suggestion
Copyright 2026 The Kubernetes Authors.
```

richardcase commented on test/e2e/suites/managed/eks_scaling_config_test.go:
```diff
@@ -0,0 +1,146 @@
+		clusterName string
+	)
+
+	ginkgo.It("[managed] [scaling-config] Able to create cluster with scaling tier and update from standard to tier-xl", func() {
```
What are the cost implications of provisioned control plane?

If there is a significant price difference to non-provisioned we may want to mark this is "pending" and then when we have the weekly "full" suite re-enable the test.

cespo replied:
The price difference is quite steep. You go from ~$70 of standard to ~$1200 for XL and up to ~$10k for a 8XL. I will mark the test as pending so it won't run in the regular CI suite

cespo committed:
Add support for EKS Provisioned Control Plane

This change adds support for the EKS Provisioned Control Plane feature,
which allows users to select from a set of scaling tiers (standard,
tier-xl, tier-2xl, tier-4xl) for predictable, high performance from
the cluster's control plane.

Changes:
- Add ControlPlaneScalingConfig field to AWSManagedControlPlane API
- Add ControlPlaneScalingTier type with kubebuilder default and validation
- Add Equal method to ControlPlaneScalingConfig for explicit comparison
- Implement converters between CAPA and AWS SDK types
- Update cluster creation to include scaling configuration
- Add reconciliation logic for updating scaling configuration
- Update v1beta1 conversion functions for backward compatibility
- Upgrade AWS SDK for EKS to v1.75.1 to support the new API
- Add unit tests and e2e test (marked Pending due to cost)

Signed-off-by: Carmine Esposito <cesposito@atlassian.com>

cespo commented:
/retest

richardcase commented:
/test pull-cluster-api-provider-aws-e2e-eks

richardcase commented:
/test pull-cluster-api-provider-aws-e2e-eks

cespo commented:
/test pull-cluster-api-provider-aws-e2e-eks

cespo commented:
/test pull-cluster-api-provider-aws-e2e-eks

richardcase commented:
@cespo - sorry its taken me so long to get back to this. Any chance you could rebase?

richardcase commented:
/test pull-cluster-api-provider-aws-e2e-eks

richardcase commented:
/test pull-cluster-api-provider-aws-e2e-eks

cespo commented:
/retest

cespo commented:
> @cespo - sorry its taken me so long to get back to this. Any chance you could rebase?

I rebased but the tests are still failing

cespo commented:
/test pull-cluster-api-provider-aws-e2e-eks

richardcase commented:
/approve

richardcase commented:
/lgtm
